analyzers/src/SonarAnalyzer.VisualBasic/Rules/UseReturnStatement.cs

```diff
@@ -0,0 +1,86 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.VisualBasic;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.VisualBasic
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [Rule(DiagnosticId)]
+    public sealed class UseReturnStatement : SonarDiagnosticAnalyzer
+    {
+        private const string DiagnosticId = "S5944";
+        private const string MessageFormat = "{0}";
+        private const string ReturnMessage = "Use a 'Return' statement; assigning returned values to function names is obsolete.";
+        private const string UsageMessage = "Do not make use of the implicit return value.";
+
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var method = (MethodStatementSyntax)((MethodBlockSyntax)c.Node).BlockStatement;
+                new IdentifierWalker(c, method.Identifier.ValueText).SafeVisit(c.Node);
+            },
+            SyntaxKind.FunctionBlock);
+
+        private class IdentifierWalker : VisualBasicSyntaxWalker
+        {
+            private readonly SyntaxNodeAnalysisContext context;
+            private readonly string name;
+
+            public IdentifierWalker(SyntaxNodeAnalysisContext context, string name)
+            {
+                this.context = context;
+                this.name = name;
+            }
+
+            public override void VisitIdentifierName(IdentifierNameSyntax node)
+            {
+                if (IsImplicitReturnValue(node))
+                {
+                    context.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, node.GetLocation(), NoAssignment(node) ? UsageMessage : ReturnMessage));
+                }
+            }
+
+            private bool IsImplicitReturnValue(IdentifierNameSyntax node) =>
```

Comment:
I'd embed this one now :) As the logic is easy to read and it leads directly to reporting the issue.